src/goto-instrument/contracts/dynamic-frames/dfcc_instrument_loop.cpp

```diff
@@ -0,0 +1,703 @@
+  //   havoc(assigns_clause_targets);
+  //   in_loop_havoc_block = false;
+  //   ASSUME(loop_invariant);
+  //   old_variant = loop_decreases;
+  // ```
+
+  goto_programt step_instrs;
+  goto_programt::instructiont::targett step_case_target;
+
+  {
+    // We skip past it initially, because of the unconditional jump above,
+    // but jump back here if we get past the loop guard while in_base_case.
+    // `in_base_case = false;`
+    step_case_target = step_instrs.add(goto_programt::make_assignment(
+      in_base_case, false_exprt{}, loop_head_location));
+  }
+
+  {
+    // If we jump here, then the loop runs at least once,
+    // so add the base case assertion: `assert(initial_invariant)`.
+    source_locationt check_location(loop_head_location);
+    check_location.set_property_class("loop_invariant_base");
+    check_location.set_comment(
+      "Check invariant before entry for loop " +
+      id2string(check_location.get_function()) + "." +
+      std::to_string(cbmc_loop_id));
+    step_instrs.add(
+      goto_programt::make_assertion(initial_invariant, check_location));
+  }
+
+  {
+    // Check assigns clause inclusion with parent write set
+    // skip the check when if w_parent is NULL.
+    auto goto_instruction = step_instrs.add(goto_programt::make_incomplete_goto(
+      equal_exprt(
+        outer_write_set,
+        null_pointer_exprt(to_pointer_type(outer_write_set.type()))),
+      loop_head_location));
+
+    auto check_var =
+      get_fresh_aux_symbol(
+        bool_typet(),
+        id2string(function_id),
+        "__check_assigns_clause_incl_loop_" + std::to_string(cbmc_loop_id),
+        loop_head_location,
+        language_mode,
+        symbol_table)
+        .symbol_expr();
+
+    step_instrs.add(goto_programt::make_decl(check_var, loop_head_location));
+    step_instrs.add(goto_programt::make_function_call(
+      library.write_set_check_assigns_clause_inclusion_call(
+        check_var, outer_write_set, addr_of_loop_write_set, loop_head_location),
+      loop_head_location));
+
+    source_locationt check_location(loop_head_location);
+    check_location.set_property_class("loop_assigns");
+    check_location.set_comment(
+      "Check assigns clause inclusion for loop " +
+      id2string(check_location.get_function()) + "." +
+      std::to_string(cbmc_loop_id));
+    step_instrs.add(goto_programt::make_assertion(check_var, check_location));
+    step_instrs.add(goto_programt::make_dead(check_var, loop_head_location));
+
+    auto label_instruction =
+      step_instrs.add(goto_programt::make_skip(loop_head_location));
+    goto_instruction->complete_goto(label_instruction);
+  }
+
+  {
+    // Generate havocing code for assigns targets.
+    const auto in_loop_havoc_block =
+      new_tmp_symbol(
+        bool_typet(),
+        loop_head_location,
+        language_mode,
+        symbol_table,
+        std::string(IN_LOOP_HAVOC_BLOCK) + +"__" + std::to_string(cbmc_loop_id))
+        .symbol_expr();
```

Comment:
When do you use `new_tmp_symbol` vs directly using `get_fresh_aux_symbol`? (I'd actually consider getting rid of `new_tmp_symbol` altogether...)?

Reply:
All `new_tmp_symbol` does is extract function id from loop location and set the symbol auxiliary. However, it calls `get_fresh_aux_symbol` which already return auxiliary symbol.
So yes, we should get rid of `new_tmp_symbol`.